tools/config/main.c

```diff
@@ -0,0 +1,477 @@
+        if (cn_root == NULL)
+            fail(ERR_SYSTEM, "could not parse config file %s\n", main_config);
+        if (pfilename)
+            *pfilename = strdup(main_config);
+    } else {
+        char *repodir = get_repodir(main_config);
+
+        cn_root = find_repo(repodir, repo, pfilename);
+        safe_free(repodir);
+        if (cn_root == NULL)
+            fail(ERR_NO_REPO, "repo '%s' not found\n", repo);
+    }
+    return cn_root;
+}
+
+static
+int write_file(struct cnfnode *cn_root, const char *filename)
+{
+    int rc = 0;
+    char buf[256];
+
+    snprintf(buf, sizeof(buf), "%s.tmp", filename);
+    rc = cnfmodule_unparse_file(mod_ini, buf, cn_root);
+    check_cond(rc == 0);
+
+    rc = rename(buf, filename);
+    check_cond(rc == 0);
+error:
+    return rc;
+}
+
+static
+struct json_dump *cnftree2json(struct cnfnode *cn_root)
+{
+    int rc = 0;
+    struct json_dump *jd = jd_create(0);
+
+    jd_map_start(jd);
+    if (cn_root->first_child) {
+        struct json_dump *jd_child = cnftree2json(cn_root->first_child);
+        check_ptr(jd_child);
+        rc = jd_map_add_child(jd, cn_root->name, jd_child);
+        check_rc(rc);
+        jd_destroy(jd_child);
+    } else {
+        for (struct cnfnode *cn = cn_root; cn; cn = cn->next) {
+            rc = jd_map_add_string(jd, cn->name, cn->value);
+            check_rc(rc);
+        }
+    }
+error:
+    if (rc && jd){
+        jd_destroy(jd);
+        jd = NULL;
+    }
+    return jd;
+}
+
+int main(int argc, char *argv[])
+{
+    char *main_config = TDNF_CONF_FILE;
+    char *repo_config = NULL;
+    int do_json = 0;
+    int rc = 0;
+
+    while(1) {
+        int c;
+
+        static struct option long_options[] = {
+            {"config", 1, 0, 'c'},
+            {"file", 1, 0, 'f'},
+            {"json", 1, 0, 'j'},
+            {0, 0, 0, 0}
+        };
+
+        c = getopt_long(argc, argv, "c:f:j",
+            long_options, NULL);
+
+        if (c == -1)
+            break;
+
+        switch(c){
+        case 'c':
+            main_config = optarg;
+            break;
+        case 'f':
+            repo_config = optarg;
+            break;
+        case 'j':
+            do_json = 1;
+            break;
+        case '?':
+        default:
+            /* If it's an error, getopt has already produced an error message. */
+            //usage();
+            return 1;
+        }
+    }
+
+    register_ini(NULL);
+    mod_ini = find_cnfmodule("ini");
+
+    /*
+     * Process the action(s).
+     */
+    if (optind < argc) {
+        int argcount = 0;
+        char *action = NULL;
+        char *repo = NULL;
+        char *filename = NULL;
+        struct cnfnode *cn_root = NULL;
+
+        while (optind + argcount < argc)
+            argcount++;
+
+       	/*
+        * Find the action.
+        */
+        action = argv[optind];
+        if(strcmp(action, "edit") == 0 || strcmp(action, "create") == 0) {
```

Comment:
can we have all option like edit, create, modify as a macros

Reply:
I don't like that. It makes it harder to read because then we'd need to look up how those macros are defined.